### DISCUSSION TITLE
CUDA OOM during integrate_tsdf_with_features with high-resolution dataset

### DISCUSSION THREAD

**Discussion by lorenzobagnol** [openvdb/fvdb-core#660] [Q&A]
Hi everyone,

I'm using fVDB Reality Capture with fVDB-core v. 0.4.2 to reconstruct very complex meshes where I need to preserve a very high level of geometric detail.

My dataset consists of:

~600 high-quality images
Resolution: 6000×4000
Running on a cloud instance with an NVIDIA L4 GPU (24 GB VRAM)

To preserve reconstruction quality, I need to keep both:

a small voxel_size
a small truncation_margin

However, during the call to integrate_tsdf_with_features, I consistently hit a CUDA Out-Of-Memory (OOM) error.

Below is the relevant integration loop:
       
        enumerator = tqdm.tqdm(dataloader, unit="imgs", desc="Extracting TSDF") if show_progress else dataloader

        for i, tsdf_input in enumerate(enumerator):
            cam_to_world_matrix = camera_to_world_matrices[i].to(dtype=torch.float32, device=device)
            projection_matrix = projection_matrices[i].to(dtype=torch.float32, device=device)

            if len(tsdf_input) == 4:
                rgb_image, depth_image, weight_image, mask_image = tsdf_input
            else:
                rgb_image, depth_image, weight_image = tsdf_input
                mask_image = None

            # Convert rgb image to feature dtype
            if feature_dtype == torch.uint8:
                rgb_image = (rgb_image * 255).to(feature_dtype)
            else:
                rgb_image = rgb_image.to(feature_dtype)

            # Prepare depth and weight tensors
            depth_image = depth_image.to(dtype)
            weight_image = weight_image.to(dtype)

            # Prepare mask feature channels (if provided) in the same feature dtype.
            # mask_image from __getitem__ is (N, H, W); DataLoader wraps it to (B, N, H, W).
            mask_feat = None
            if mask_image is not None:
                # Normalise to (B, N, H, W) — handle both batched and unbatched cases
                if mask_image.dim() == 3:
                    mask_image = mask_image.unsqueeze(0)   # (N, H, W) → (1, N, H, W)
                # (B, N, H, W) → (B, H, W, N) channel-last for concatenation with rgb
                mask_image = mask_image.permute(0, 2, 3, 1).contiguous()

                if feature_dtype == torch.uint8:
                    mask_feat = (mask_image * 255).clamp(0, 255).to(feature_dtype)
                else:
                    mask_feat = mask_image.to(feature_dtype)

                mask_image = mask_image.to(dtype)

            # Build the full feature tensor (B, H, W, 3+N) for integration
            if features.shape[1] > model.num_channels:
                mask_ch = features.shape[1] - model.num_channels   # == mask_channel_count
                if mask_feat is None:
                    zeros_mask = torch.zeros(
                        (rgb_image.shape[0], rgb_image.shape[1], rgb_image.shape[2], mask_ch),
                        device=rgb_image.device,
                        dtype=feature_dtype,
                    )
                    rgb_for_integration = torch.cat([rgb_image, zeros_mask], dim=-1)
                else:
                    # Pad mask_feat if it has fewer channels than reserved (safety guard)
                    if mask_feat.shape[-1] < mask_ch:
                        pad = torch.zeros(
                            (*mask_feat.shape[:-1], mask_ch - mask_feat.shape[-1]),
                            device=mask_feat.device,
                            dtype=feature_dtype,
                        )
                        mask_feat = torch.cat([mask_feat, pad], dim=-1)
                    rgb_for_integration = torch.cat([rgb_image, mask_feat], dim=-1)
            else:
                rgb_for_integration = rgb_image

            accum_grid, tsdf, weights, features = accum_grid.integrate_tsdf_with_features(
                truncation_margin,
                projection_matrix.to(dtype),
                cam_to_world_matrix.to(dtype),
                tsdf,
                features,
                weights,
                depth_image.squeeze(0).to(device),
                rgb_for_integration.squeeze(0).to(device),
                weight_image.squeeze(0).to(device),
            )

            if show_progress:
                assert isinstance(enumerator, tqdm.tqdm)
                enumerator.set_postfix({"accumulated_voxels": accum_grid.num_voxels})

            # Prune out zero weight voxels to save memory
            new_grid = accum_grid.pruned_grid(weights > 0.0)
            tsdf = new_grid.inject_from(accum_grid, tsdf)
            features = new_grid.inject_from(accum_grid, features)
            weights = new_grid.inject_from(accum_grid, weights)
            accum_grid = new_grid

            del rgb_image, depth_image, weight_image, mask_image
            torch.cuda.synchronize()
            torch.cuda.empty_cache()

**Questions**
Is this level of memory usage expected for this function with high-resolution images and dense voxel grids?
Are there recommended strategies to reduce VRAM usage while preserving reconstruction quality?
Would more aggressive voxel pruning help?
Right now I only prune voxels with weights > 0.0
Is there a way to not load everything on GPU?

At the moment it feels like memory usage grows excessively fast when using small voxel sizes and truncation margins.

Any suggestions would be greatly appreciated.

**Comment by harrism** [comment]
Closing as duplicate of #666
